### ISSUE TITLE
test.py: investigate why topology tests are slow

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#12281]
`test_topology` takes some time to run. I'm pretty sure there's something that we could do more efficiently.

I did some tests on my laptop:
- with `--unsafe-bypass-fsync 1`, running Scyllas in memory (`--tmpdir /tmp/testlog/`): 104s
- with `--unsafe-bypass-fsync 1`, running on disk: 123s
- without `--unsafe-bypass-fsync 1`, running on disk: 166s

It's curious that running in memory only decreases the runtime by ~20s.

Some observations (from looking at the logs):
- there are hangs when waiting for schema agreement. Known issue in the driver (https://github.com/scylladb/python-driver/issues/168), but perhaps we could improve the situation on test side too (e.g. decrease the schema agreement timeout, or don't wait for schema agreement in some situations when we know it's impossible anyway)
- I think there's a lot of waiting for clusters to start (before the test itself starts). Is it just that cluster startup is slow (if so, why? even with `ring_delay=0` and `skip_wait_for_gossip_to_settle`), or there's some inefficiency in how cluster pool is implemented? 
- The test cases in a single file are being run sequentially. Perhaps we could run them concurrently. It doesn't seem that `pytest` supports this out of the box, we'd need to implement this manually, e.g. run `pytest` once just to discover the list of test cases, then run separate `pytest` processes to run the test (but make sure they all use the same cluster pool). Also the concurrent runs would require separate log files (which would be good for the developer, actually - now we have a single file like `test_topology.1.log` for all test cases)

**Comment by kbr-scylla** [comment]
cc @nyh

**Comment by kostja** [comment]
Sorry, I think this is too broad and not appropriate for the issue tracker.

**Comment by kbr-scylla** [comment]
I disagree, look at issue https://github.com/scylladb/scylladb/issues/1.

**Comment by nyh** [comment]
I agree with @kbr-scylla that it's useful to have issues open even when there might not be a single specific patch that closes them. If later @kbr-scylla sends a patch to improve the performance of topology tests, he can add a "Refs" with this issue to explain to the reviewers why he's doing that patch - it's better than a patch that doesn't refer to any issue at all.

Regarding the specific points that Kamil raised:
1. I believe that use-run tests should always use /tmp, and this is what `test/cql-pytest/run.py` does. If Jenkins wants to runs tests on real disks because it believe just a test is more "real", then so be it, but developers don't need to this slowdown.
2. I don't believe in excessive parallelization. It's a good way to hide ridiculous sleeps, but it's even better to hunt down and eliminate these ridiculous sleeps. In any case, you already have parallelization in test.py, of the separate test files, I think this is more than enough. Maybe not now when you have just 3 files, but eventually you'll have many many more. cql-pytest has around 70 test files (and around 900 individual test functions).
3. I agree that booting Scylla quickly is very important (test/cql-pytest/run.py boots a single-node Scylla in half a second) and you should definitely use whatever options - however hacky - that allow you to boot multiple nodes quick.y But I would also look into how the test framework can recycle already-up instances of Scylla without restarting them at all. For example, if several tests need a 3-node cluster, you can reuse it. Moreover, if a test starts with a 3-node cluster and restarts one of the node, it ends with a 3-node cluster so that can be reused as well instead of being declared "dirty" and stopped.
